File: android/src/main/java/es/opendash/protocol/Frame.kt
package es.opendash.protocol

/**
 * Message framing.
 *
 * ```
 * 0..1   seq    u16 LE   TCP: always 0xffff
 * 2..3   token  u16 LE   TCP: always 0
 * 4..7   h4     u32 LE   fingerprint, checked by the firmware
 * 8..11  h8     u32 LE   context / handle
 * 12     op
 * 13..14 len    u16 LE
 * 15     chk    (op + len_lo + len_hi + 0x55) and 0xff
 * 16..   data
 * ```
 *
 * The detail that costs a session if missed: a message occupies
 * `align(16 + len, 4)` bytes on the wire, not `16 + len`. Send the short form
 * and the device waits for bytes that never arrive, the stream desynchronises
 * and the connection is reset. Measured over 4066 messages from two
 * independent captures.
 */
object Frame {

    const val HEADER = 16

    fun stride(len: Int): Int = (HEADER + len + 3) and 3.inv()

    fun checksum(op: Int, len: Int): Int =
        (op + (len and 0xff) + ((len ushr 8) and 0xff) + 0x55) and 0xff

    class Message(
        val seq: Int,
        val h4: Int,
        val h8: Int,
        val op: Int,
        val data: ByteArray,
    )

    /** Parse one message at [offset]. Returns it and how many bytes it took. */
    fun decode(buf: ByteArray, offset: Int = 0): Pair<Message, Int>? {
        if (offset + HEADER > buf.size) return null
        val op = buf[offset + 12].toInt() and 0xff
        val len = (buf[offset + 13].toInt() and 0xff) or ((buf[offset + 14].toInt() and 0xff) shl 8)
        if ((buf[offset + 15].toInt() and 0xff) != checksum(op, len)) return null
        if (offset + HEADER + len > buf.size) return null
        val msg = Message(
            seq = le16(buf, offset),
            h4 = le32(buf, offset + 4),
            h8 = le32(buf, offset + 8),
            op = op,
            data = buf.copyOfRange(offset + HEADER, offset + HEADER + len),
        )
        return msg to minOf(stride(len), buf.size - offset)
    }

    /** Walk a buffer, skipping anything that does not parse. */
    fun split(buf: ByteArray): List<Message> {
        val out = ArrayList<Message>()
        var off = 0
        while (off + HEADER <= buf.size) {
            val r = decode(buf, off)
            if (r == null) {
                off++
            } else {
                out.add(r.first)
                off += r.second
            }
        }
        return out
    }

    fun le16(b: ByteArray, o: Int): Int =
        (b[o].toInt() and 0xff) or ((b[o + 1].toInt() and 0xff) shl 8)

    fun le32(b: ByteArray, o: Int): Int =
        (b[o].toInt() and 0xff) or ((b[o + 1].toInt() and 0xff) shl 8) or
            ((b[o + 2].toInt() and 0xff) shl 16) or ((b[o + 3].toInt() and 0xff) shl 24)

    fun putLe32(b: ByteArray, o: Int, v: Int) {
        b[o] = (v and 0xff).toByte()
        b[o + 1] = ((v ushr 8) and 0xff).toByte()
        b[o + 2] = ((v ushr 16) and 0xff).toByte()
        b[o + 3] = ((v ushr 24) and 0xff).toByte()
    }
}

File: core/src/frame.rs
//! Message framing.
//!
//! Every message is a 16-byte header followed by `len` bytes of data:
//!
//! ```text
//! 0..1   seq    u16 LE   USB: a counter · TCP: always 0xffff
//! 2..3   token  u16 LE   USB: the driver's echo · TCP: always 0
//! 4..7   h4     u32 LE   fingerprint over seq, h8, the data and the padding
//! 8..11  h8     u32 LE   context / handle
//! 12     op
//! 13..14 len    u16 LE
//! 15     chk    (op + len_lo + len_hi + 0x55) & 0xff
//! 16..   data
//! ```
//!
//! What matters and is easy to get wrong: a message does **not** occupy
//! `16 + len` bytes on the wire. It occupies that rounded up to a multiple of
//! four. Send the short version and the device waits for the missing bytes,
//! the stream desynchronises and the connection is reset. Measured over 4066
//! messages from two independent sessions: the gap between one message and the
//! next is always exactly this padding.

pub const HEADER: usize = 16;

/// Bytes a message really occupies, padding included.
pub const fn stride(len: usize) -> usize {
    (HEADER + len + 3) & !3
}

pub const fn checksum(op: u8, len: u16) -> u8 {
    (op as u16)
        .wrapping_add(len & 0xff)
        .wrapping_add((len >> 8) & 0xff)
        .wrapping_add(0x55) as u8
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Message {
    pub seq: u16,
    pub token: u16,
    pub h4: u32,
    pub h8: u32,
    pub op: u8,
    pub data: Vec<u8>,
}

impl Message {
    /// Serialise, padding to the four-byte stride. The padding the device
    /// sends is stale buffer content, not zeros, but its value does not
    /// matter — only that the bytes are there.
    pub fn encode(&self) -> Vec<u8> {
        let len = self.data.len();
        let mut out = vec![0u8; stride(len)];
        out[0..2].copy_from_slice(&self.seq.to_le_bytes());
        out[2..4].copy_from_slice(&self.token.to_le_bytes());
        out[4..8].copy_from_slice(&self.h4.to_le_bytes());
        out[8..12].copy_from_slice(&self.h8.to_le_bytes());
        out[12] = self.op;
        out[13..15].copy_from_slice(&(len as u16).to_le_bytes());
        out[15] = checksum(self.op, len as u16);
        out[HEADER..HEADER + len].copy_from_slice(&self.data);
        out
    }

    /// Parse one message at the start of `buf`. Returns it with the number of
    /// bytes consumed, which is the padded stride.
    pub fn decode(buf: &[u8]) -> Option<(Message, usize)> {
        if buf.len() < HEADER {
            return None;
        }
        let op = buf[12];
        let len = u16::from_le_bytes([buf[13], buf[14]]);
        if buf[15] != checksum(op, len) {
            return None;
        }
        let len = len as usize;
        if buf.len() < HEADER + len {
            return None;
        }
        let msg = Message {
            seq: u16::from_le_bytes([buf[0], buf[1]]),
            token: u16::from_le_bytes([buf[2], buf[3]]),
            h4: u32::from_le_bytes([buf[4], buf[5], buf[6], buf[7]]),
            h8: u32::from_le_bytes([buf[8], buf[9], buf[10], buf[11]]),
            op,
            data: buf[HEADER..HEADER + len].to_vec(),
        };
        Some((msg, stride(len).min(buf.len())))
    }
}

/// Walk a stream, skipping anything that does not parse.
pub fn split(buf: &[u8]) -> Vec<Message> {
    let mut out = Vec::new();
    let mut off = 0;
    while off + HEADER <= buf.len() {
        match Message::decode(&buf[off..]) {
            Some((m, used)) => {
                off += used;
                out.push(m);
            }
            None => off += 1,
        }
    }
    out
}
